### PR TITLE
Replace bind expr with stage mem

### PR DESCRIPTION
**Pull request by SamirDroubi**
An attempt at addressing #160 

It removes all usages of `bind_expr` and replaces it with `stage_mem` or a new stdlib function `stage_expr`. Only remaining usages are in examples/demo.

`bind_expr` is used throughout the codebase to perform manual staging of expressions. Many of those staging patterns can be refactored to completely use `stage_mem`. However, there are a few situations where that's not possible.
1. Staging to do a broadcast (for example when you want to stage `A` in matrix multiply, see test_x86.py::avx2_sgemm_6x16).
3. Staging a binop expressions (see test_x86.py::simple_math_avx2_sched)
4. Staging the temporary buffer into a buffer with higher dimensions than the original buffer (see test_im2col.py::test_im2col)

1 and 2 code patterns could be refactored into one function which I added into stdlib called `stage_expr`. However, the function ended up being too parameterized and has a problem where it could accept multiple expression cursors in the first step of `bind_expr`, but the function is really assuming one expression cursor. 

For 3, I simply replaced the call to `bind_expr` with a call to `stage_mem` but left the remainder of the staging pattern.

It seems that it is not possible to completely deprecate `bind_expr` because it is necessary for usages of binop expressions staging as in 2. Moreover, it seems that the implementation of `stage_expr` is not a good idea at least in it is the current way it is written. Finally, for 3 it seems more natural to use `bind_expr` instead of `stage_mem` since in that case, `stage_mem` is behaving exactly like `bind_expr` by removing the store subsequent to the block. For cases like 1, we could try to remove the staging pattern with `stage_mem` and try to use other scheduling primitives to achieve broadcasting, but that seems unnecessary if we have to keep `bind_expr` anyways.

In short, it might be best to keep `bind_expr`, and only replace its usage with `stage_mem` where the entire staging pattern could be removed.

### CODE REVIEW

**Comment by SamirDroubi** [comment]
`stage_expr` doesn't seem like a good addition to stdlib. It has some problems as outlined above. `bind_expr` is useful to keep around.
